### DISCUSSION TITLE
Unable to Detect Camera over Jetson Nano

### DISCUSSION THREAD

**Discussion by arpit6232** [OE4T/meta-tegra#653] [Q&A] [answered]
Environment Details : 
Board : Jetson Nano ( Developer Kit )
Branch : `dunfell-l4t-r32.4.3`
Commit Hash : `d5348114b0a9acd42a473aa4542a19f27f3a2a2c`
Camera Connected #1 : Logitech C270
Camera Connected #2 : Raspberry Pi Camera V2.1
Image : `demo-image-full`
Yocto Setup : `. ./setup-env --machine jetson-nano-qspi-sd`

The intention was to use either Logitech C270 Camera or Raspberry Pi Camera using gstreamer for image processing. I ran into this issue where the Image built does not recognize either of the camera. The Camera are functioning correctly and were tested on NVIDIA Image binaries on the official website  [JP 4.3 version](https://developer.nvidia.com/embedded/downloads), but also failed on [JP 4.5.1](https://developer.nvidia.com/embedded/jetpack). 

Below was found on the Yocto build for the Jetson Nano. I have attached the `dmesg` on the Jetson Nano , with and without the camera attached. The camera are not detected under `/dev/ ` which should be `/dev/video` to be used by gstreamer. 

![IMG_20210407_173817__01](https://user-images.githubusercontent.com/55374040/113948521-4adb3500-97ca-11eb-8352-fc283799faab.jpg)

**Comment by dwalkes** [comment]
@arpit6232 since you've noticed this issue with stock nvidia jetpack you might want to cross post on the NVIDIA developer forum

**Reply by arpit6232** [reply]
Absolutely.

**Comment by ichergui** [comment]
Hey @arpit6232 

Nvidia provides additional TARBALL that contains (binaries, library and DTB ) from `Jetpack 4.4.1` / `L4T-R32.4.4`, this is specific for RPI HQ Camera (`IMX477 sensor support`). Please check [here](https://developer.nvidia.com/embedded/linux-tegra-r3244).

Here is the [RPI camera HQ](https://www.waveshare.com/raspberry-pi-hq-camera.htm)

I'm not sure about the connection between the camera (`Logitech C270`) and the Jetson Nano, maybe you are using USB, right ? 
Concerning the RPI camera, I assume that is connected through `CSI`. 

Please check the kernel config file and enable the drivers needed, you may have to update the dtb as well. I have the [RPI camera v2](https://www.raspberrypi.org/products/camera-module-v2/) that I can try it later with my Jetson Nano. I will let you know.

Here is the [RPI cameras](https://www.raspberrypi.org/documentation/hardware/camera/) 

Could you please share the one that you are using ?

**Reply by arpit6232** [reply]
I am using the CSI module for the [raspberry Pi camera](https://www.raspberrypi.org/documentation/hardware/camera/) . 
Upon little tweaks as you suggested I got the Raspberry Pi Camera setup showing under /dev, But could not see anything for the USB Camera.

**Reply by ichergui** [reply]
Please see @madisongh suggestion

**Comment by madisongh** [comment, marked as answer]
Probably best to start with making sure the kernel is recognizing the hardware and the appropriate drivers are getting loaded. Can you post the `dmesg` or `journalctl -k -b`  output?  Does the Logitech show up in `lsusb` output?

**Reply by arpit6232** [reply]
The USB Camera gets detected under `lsusb` . The dmesg output looks like in the image below. But does not get recognized under /dev/video . I checked for if there were any udev issues, But it seemed to have to located correctly. 

[image](https://user-images.githubusercontent.com/55374040/114068720-6f332200-985b-11eb-8df6-e723a5c644c7.png)

**Reply by madisongh** [reply]
Are you adding `kernel-module-uvcvideo` to your build?  I have a C930e which has worked fine for me, but since the uvcvideo driver is built as a module, rather than into the kernel, I have to explicitly add it to my images.

**Reply by arpit6232** [reply]
I believe that is part of openEmbedded layer. As per the below image. 
![image](https://user-images.githubusercontent.com/55374040/114226955-90664200-9931-11eb-876f-d7644cc2e9ee.png)

Despite that I added the `kernel-module-uvcvideo` but it still does not seem to detect it USB Camera (Logitech C270).

**Reply by madisongh** [reply]
So check `lsmod` output on the Jetson to see if the module got loaded into the kernel, and look at the kernel log messages (from `dmesg` or `journalctl -k -b`) to look for the USB bus probing, to see if it identifies the device, and to see if there is any output from the uvcvideo driver.
